File: quickshell/daily-network-logger.py
#!/usr/bin/env python3
"""
Daily Internet Data Consumption Tracker by Wi-Fi Network & Physical Interfaces.
Logs daily statistics to monthly Markdown tables in ~/Data Consumption/YYYY-MM.md.
"""

import os
import sys
import glob
import json
import time
import subprocess
from datetime import datetime

HOME_DIR = os.path.expanduser("~")
CONSUMPTION_DIR = os.path.join(HOME_DIR, "Data Consumption")
STATE_DIR = os.path.join(HOME_DIR, ".local", "share", "quickshell")
STATE_FILE = os.path.join(STATE_DIR, "network_logger_state.json")

def get_boot_id():
    try:
        with open("/proc/sys/kernel/random/boot_id", "r") as f:
            return f.read().strip()
    except Exception:
        return "unknown"

def fmt_bytes(b):
    if b >= 1024**3:
        # 3 decimal places for GB (e.g., 1.234 GB)
        return f"{b / (1024**3):.3f} GB"
    elif b >= 1024**2:
        return f"{b / (1024**2):.1f} MB"
    elif b >= 1024:
        return f"{b / 1024:.0f} KB"
    else:
        return f"{b} B"

def is_physical_interface(dev_path):
    iface = os.path.basename(dev_path)
    if iface.startswith(("lo", "docker", "veth", "virbr", "br-", "tun", "wg", "tailscale", "dummy", "zt")):
        return False
    if os.path.exists(os.path.join(dev_path, "device")):
        return True
    if iface.startswith(("wlan", "wlp", "enp", "eth", "eno", "usb", "enx", "wwan", "wwp")):
        return True
    return False

def get_active_network_for_interface(iface):
    """Detects SSID or connection name for a specific interface."""
    try:
        proc = subprocess.run(
            ["nmcli", "-t", "-f", "DEVICE,TYPE,STATE,CONNECTION", "dev"],
            capture_output=True, text=True, timeout=2
        )
        for line in proc.stdout.strip().split("\n"):
            if not line:
                continue
            parts = line.split(":")
            if len(parts) >= 4 and parts[0] == iface and "connected" in parts[2]:
                conn_name = parts[3].strip()
                dev_type = parts[1].strip().lower()
                if conn_name:
                    return conn_name
                if dev_type == "wifi":
                    return "Wi-Fi"
                elif dev_type == "ethernet":
                    return "Ethernet"
    except Exception:
        pass
    
    # Fallback heuristic based on interface name
    if iface.startswith(("wlan", "wlp")):
        try:
            ssid_proc = subprocess.run(["iwgetid", "-r"], capture_output=True, text=True, timeout=1)
            ssid = ssid_proc.stdout.strip()
            if ssid:
                return ssid
        except Exception:
            pass
        return "Wi-Fi"
    elif iface.startswith(("usb", "enx")):
        return "USB Tethering"
    elif iface.startswith(("enp", "eth", "eno")):
        return "Ethernet"
    elif iface.startswith(("wwan", "wwp")):
        return "Cellular"
    return "Network"

def load_state():
    if os.path.exists(STATE_FILE):
        try:
            with open(STATE_FILE, "r") as f:
                return json.load(f)
        except Exception:
            pass
    return {
        "boot_id": get_boot_id(),
        "iface_stats": {},
        "records": {}
    }

def save_state(state):
    os.makedirs(STATE_DIR, exist_ok=True)
    try:
        with open(STATE_FILE + ".tmp", "w") as f:
            json.dump(state, f, indent=2)
        os.replace(STATE_FILE + ".tmp", STATE_FILE)
    except Exception:
        pass

def update_data_consumption():
    os.makedirs(CONSUMPTION_DIR, exist_ok=True)
    now = datetime.now()
    cur_month = now.strftime("%Y-%m")
    cur_date = now.strftime("%Y-%m-%d")
    cur_boot_id = get_boot_id()

    state = load_state()
    saved_boot_id = state.get("boot_id", "")
    is_new_boot = (cur_boot_id != saved_boot_id)
    saved_ifaces = state.get("iface_stats", {})
    records = state.get("records", {})

    if cur_month not in records:
        records[cur_month] = {}
    if cur_date not in records[cur_month]:
        records[cur_month][cur_date] = {}

    for dev_path in glob.glob("/sys/class/net/*"):
        if not is_physical_interface(dev_path):
            continue
        iface = os.path.basename(dev_path)
        try:
            with open(os.path.join(dev_path, "statistics/rx_bytes")) as f:
                rx = int(f.read().strip())
            with open(os.path.join(dev_path, "statistics/tx_bytes")) as f:
                tx = int(f.read().strip())
            
            if iface not in saved_ifaces:
                # Newly detected interface: initialize baseline
                saved_ifaces[iface] = {"rx": rx, "tx": tx}
                delta_rx = 0
                delta_tx = 0
            elif is_new_boot:
                delta_rx = rx
                delta_tx = tx
                saved_ifaces[iface] = {"rx": rx, "tx": tx}
            else:
                prev = saved_ifaces[iface]
                delta_rx = rx - prev.get("rx", 0) if rx >= prev.get("rx", 0) else rx
                delta_tx = tx - prev.get("tx", 0) if tx >= prev.get("tx", 0) else tx
                saved_ifaces[iface] = {"rx": rx, "tx": tx}

            if delta_rx > 0 or delta_tx > 0:
                net_name = get_active_network_for_interface(iface)
                if net_name not in records[cur_month][cur_date]:
                    records[cur_month][cur_date][net_name] = {"rx": 0, "tx": 0}
                records[cur_month][cur_date][net_name]["rx"] += delta_rx
                records[cur_month][cur_date][net_name]["tx"] += delta_tx
        except Exception:
            pass

    state["boot_id"] = cur_boot_id
    state["iface_stats"] = saved_ifaces
    state["records"] = records
    save_state(state)

    # Render Markdown table for the current month
    render_month_markdown(cur_month, records.get(cur_month, {}))

def render_month_markdown(month_str, month_records):
    os.makedirs(CONSUMPTION_DIR, exist_ok=True)
    md_file = os.path.join(CONSUMPTION_DIR, f"{month_str}.md")
    
    # Month Title
    try:
        month_dt = datetime.strptime(month_str, "%Y-%m")
        month_name = month_dt.strftime("%B %Y")
    except Exception:
        month_name = month_str

    lines = [
        f"# Internet Data Consumption - {month_name}",
        "",
        "| Date | Wi-Fi / Network | Downloaded | Uploaded | Total |",
        "| :--- | :--- | :--- | :--- | :--- |"
    ]

    total_month_rx = 0
    total_month_tx = 0

    # Sort dates chronologically
    sorted_dates = sorted(month_records.keys())
    for d in sorted_dates:
        networks = month_records[d]
        sorted_nets = sorted(networks.keys())
        for net in sorted_nets:
            rx = networks[net].get("rx", 0)
            tx = networks[net].get("tx", 0)
            total = rx + tx
            total_month_rx += rx
            total_month_tx += tx
            lines.append(f"| {d} | {net} | {fmt_bytes(rx)} | {fmt_bytes(tx)} | {fmt_bytes(total)} |")

    # If no data yet, provide a placeholder row
    if not sorted_dates:
        lines.append(f"| {datetime.now().strftime('%Y-%m-%d')} | No active traffic | 0 B | 0 B | 0 B |")

    lines.append("")
    lines.append("### Monthly Summary")
    lines.append(f"- **Total Downloaded**: `{fmt_bytes(total_month_rx)}`")
    lines.append(f"- **Total Uploaded**: `{fmt_bytes(total_month_tx)}`")
    lines.append(f"- **Grand Total**: `{fmt_bytes(total_month_rx + total_month_tx)}`")
    lines.append("")
    lines.append(f"*Last updated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}*")
    lines.append("")

    content = "\n".join(lines)
    try:
        with open(md_file + ".tmp", "w") as f:
            f.write(content)
        os.replace(md_file + ".tmp", md_file)
    except Exception:
        pass

def show_current_month():
    cur_month = datetime.now().strftime("%Y-%m")
    md_file = os.path.join(CONSUMPTION_DIR, f"{cur_month}.md")
    if os.path.exists(md_file):
        with open(md_file, "r") as f:
            print(f.read())
    else:
        print(f"No records found yet for {cur_month}.")

if __name__ == "__main__":
    action = sys.argv[1] if len(sys.argv) > 1 else "--update"
    if action == "--show":
        update_data_consumption()
        show_current_month()
    else:
        update_data_consumption()

File: dark-tools-rs/src/bin/daily_network_logger.rs
#[path = "../common.rs"]
mod common;

use chrono::{Local, NaiveDate};
use common::{fmt_bytes, get_boot_id, is_physical_interface};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, HashMap};
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
struct InterfaceStat {
    #[serde(default)]
    rx: u64,
    #[serde(default)]
    tx: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
struct TrafficStat {
    #[serde(default)]
    rx: u64,
    #[serde(default)]
    tx: u64,
}

#[derive(Serialize, Deserialize, Debug, Default)]
struct LoggerState {
    #[serde(default)]
    boot_id: String,
    #[serde(default)]
    iface_stats: HashMap<String, InterfaceStat>,
    #[serde(default)]
    records: BTreeMap<String, BTreeMap<String, BTreeMap<String, TrafficStat>>>,
}

fn get_home_dir() -> PathBuf {
    let home = env::var("HOME").unwrap_or_else(|_| "/home/aryan".to_string());
    PathBuf::from(home)
}

fn get_active_network_for_interface(iface: &str) -> String {
    if let Ok(out) = Command::new("nmcli")
        .args(["-t", "-f", "DEVICE,TYPE,STATE,CONNECTION", "dev"])
        .output()
    {
        if let Ok(text) = String::from_utf8(out.stdout) {
            for line in text.lines() {
                let line = line.trim();
                if line.is_empty() {
                    continue;
                }
                let parts: Vec<&str> = line.split(':').collect();
                if parts.len() >= 4 && parts[0] == iface && parts[2].contains("connected") {
                    let conn_name = parts[3].trim();
                    let dev_type = parts[1].trim().to_lowercase();
                    if !conn_name.is_empty() {
                        return conn_name.to_string();
                    }
                    if dev_type == "wifi" {
                        return "Wi-Fi".to_string();
                    } else if dev_type == "ethernet" {
                        return "Ethernet".to_string();
                    }
                }
            }
        }
    }

    // Fallback heuristic based on interface name
    if iface.starts_with("wlan") || iface.starts_with("wlp") {
        if let Ok(out) = Command::new("iwgetid").arg("-r").output() {
            if let Ok(ssid) = String::from_utf8(out.stdout) {
                let s = ssid.trim();
                if !s.is_empty() {
                    return s.to_string();
                }
            }
        }
        return "Wi-Fi".to_string();
    } else if iface.starts_with("usb") || iface.starts_with("enx") {
        return "USB Tethering".to_string();
    } else if iface.starts_with("enp") || iface.starts_with("eth") || iface.starts_with("eno") {
        return "Ethernet".to_string();
    } else if iface.starts_with("wwan") || iface.starts_with("wwp") {
        return "Cellular".to_string();
    }

    "Network".to_string()
}

fn render_month_markdown(
    month_str: &str,
    month_records: &BTreeMap<String, BTreeMap<String, TrafficStat>>,
    consumption_dir: &Path,
) {
    let _ = fs::create_dir_all(consumption_dir);
    let md_file = consumption_dir.join(format!("{}.md", month_str));

    let month_name = NaiveDate::parse_from_str(&format!("{}-01", month_str), "%Y-%m-%d")
        .map(|d| d.format("%B %Y").to_string())
        .unwrap_or_else(|_| month_str.to_string());

    let mut lines = Vec::new();
    lines.push(format!("# Internet Data Consumption - {}", month_name));
    lines.push(String::new());
    lines.push("| Date | Wi-Fi / Network | Downloaded | Uploaded | Total |".to_string());
    lines.push("| :--- | :--- | :--- | :--- | :--- |".to_string());

    let mut total_month_rx: u64 = 0;
    let mut total_month_tx: u64 = 0;

    let has_records = !month_records.is_empty();

    for (date, networks) in month_records {
        for (net, stats) in networks {
            let total = stats.rx + stats.tx;
            total_month_rx += stats.rx;
            total_month_tx += stats.tx;
            lines.push(format!(
                "| {} | {} | {} | {} | {} |",
                date,
                net,
                fmt_bytes(stats.rx),
                fmt_bytes(stats.tx),
                fmt_bytes(total)
            ));
        }
    }

    if !has_records {
        let today = Local::now().format("%Y-%m-%d").to_string();
        lines.push(format!("| {} | No active traffic | 0 B | 0 B | 0 B |", today));
    }

    lines.push(String::new());
    lines.push("### Monthly Summary".to_string());
    lines.push(format!("- **Total Downloaded**: `{}`", fmt_bytes(total_month_rx)));
    lines.push(format!("- **Total Uploaded**: `{}`", fmt_bytes(total_month_tx)));
    lines.push(format!(
        "- **Grand Total**: `{}`",
        fmt_bytes(total_month_rx + total_month_tx)
    ));
    lines.push(String::new());
    lines.push(format!(
        "*Last updated: {}*",
        Local::now().format("%Y-%m-%d %H:%M:%S")
    ));
    lines.push(String::new());

    let content = lines.join("\n");
    let tmp_file = consumption_dir.join(format!("{}.md.tmp", month_str));
    if fs::write(&tmp_file, content).is_ok() {
        let _ = fs::rename(&tmp_file, &md_file);
    }
}

fn update_data_consumption() {
    let home = get_home_dir();
    let consumption_dir = home.join("Data Consumption");
    let state_dir = home.join(".local/share/quickshell");
    let _ = fs::create_dir_all(&state_dir);
    let state_file = state_dir.join("network_logger_state.json");

    let now = Local::now();
    let cur_month = now.format("%Y-%m").to_string();
    let cur_date = now.format("%Y-%m-%d").to_string();
    let cur_boot_id = get_boot_id();

    let mut state: LoggerState = if state_file.exists() {
        fs::read_to_string(&state_file)
            .ok()
            .and_then(|s| serde_json::from_str(&s).ok())
            .unwrap_or_default()
    } else {
        LoggerState::default()
    };

    let is_new_boot = state.boot_id != cur_boot_id;

    if let Ok(entries) = fs::read_dir("/sys/class/net") {
        for entry in entries.flatten() {
            let path = entry.path();
            if !is_physical_interface(&path) {
                continue;
            }

            let iface = match path.file_name().and_then(|n| n.to_str()) {
                Some(name) => name.to_string(),
                None => continue,
            };

            let rx = fs::read_to_string(path.join("statistics/rx_bytes"))
                .ok()
                .and_then(|s| s.trim().parse::<u64>().ok())
                .unwrap_or(0);
            let tx = fs::read_to_string(path.join("statistics/tx_bytes"))
                .ok()
                .and_then(|s| s.trim().parse::<u64>().ok())
                .unwrap_or(0);

            let (delta_rx, delta_tx) = match state.iface_stats.get(&iface) {
                None => {
                    // Newly detected interface
                    state.iface_stats.insert(iface.clone(), InterfaceStat { rx, tx });
                    (0, 0)
                }
                Some(prev) => {
                    let drx = if is_new_boot {
                        rx
                    } else if rx >= prev.rx {
                        rx - prev.rx
                    } else {
                        rx
                    };
                    let dtx = if is_new_boot {
                        tx
                    } else if tx >= prev.tx {
                        tx - prev.tx
                    } else {
                        tx
                    };
                    state.iface_stats.insert(iface.clone(), InterfaceStat { rx, tx });
                    (drx, dtx)
                }
            };

            if delta_rx > 0 || delta_tx > 0 {
                let net_name = get_active_network_for_interface(&iface);
                let month_map = state.records.entry(cur_month.clone()).or_default();
                let day_map = month_map.entry(cur_date.clone()).or_default();
                let stat = day_map.entry(net_name).or_default();
                stat.rx += delta_rx;
                stat.tx += delta_tx;
            }
        }
    }

    state.boot_id = cur_boot_id;

    // Atomic write state
    let tmp_state = state_dir.join("network_logger_state.json.tmp");
    if let Ok(json_str) = serde_json::to_string_pretty(&state) {
        if fs::write(&tmp_state, json_str).is_ok() {
            let _ = fs::rename(&tmp_state, &state_file);
        }
    }

    // Render markdown table
    let empty_records = BTreeMap::new();
    let month_records = state.records.get(&cur_month).unwrap_or(&empty_records);
    render_month_markdown(&cur_month, month_records, &consumption_dir);
}

fn show_current_month() {
    let home = get_home_dir();
    let cur_month = Local::now().format("%Y-%m").to_string();
    let md_file = home.join("Data Consumption").join(format!("{}.md", cur_month));
    if let Ok(content) = fs::read_to_string(&md_file) {
        print!("{}", content);
    } else {
        println!("No records found yet for {}.", cur_month);
    }
}

fn main() {
    let action = env::args().nth(1).unwrap_or_else(|| "--update".to_string());

    if action == "--show" {
        update_data_consumption();
        show_current_month();
    } else {
        update_data_consumption();
    }
}
